File: LV 2/괄호 회전하기.py
from collections import deque

def check(s):
    st = deque()
    for t in s:
        if t in ['[', '(', '{']:
            st.append(t)
        else:
            if len(st) == 0:
                return False
            temp = st.pop()
            if (temp == '[' and t == ']') or (temp == '(' and t == ')') or (temp == '{' and t == '}'):
                continue
            return False
    if st: # 스택에 차있으면 False임! 이거 처리 안해서 13번 통과 못했었음.
        return False
    
    return True

def solution(s):
    answer = 0
    for i in range(len(s)):
        temp = s[i:] + s[:i] # 회전
        if check(temp):
            answer += 1
    return answer

File: LV 2/괄호 회전하기.java
import java.util.Optional;
import java.util.Stack;

class Solution {

    public boolean check(String s) {
        Stack<Integer> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '[' || c == '(' || c == '{') {
                stack.add((int)c);
            }
            if (stack.isEmpty()) {
                return false;
            }
            if (c == ']') {
                if (stack.peek() != '[') {
                    return false;
                }
                stack.pop();
            }
            if (c == '}') {
                if (stack.peek() != '{') {
                    return false;
                }
                stack.pop();
            }
            if (c == ')') {
                if (stack.peek() != '(') {
                    return false;
                }
                stack.pop();
            }
        }
        if (stack.isEmpty()) {
            return true;
        }
        return false;
    }

    public String rotate(String s, int count) {
        return Optional.ofNullable(s.substring(count)).orElse("")
                + Optional.ofNullable(s.substring(0, count)).orElse("");
    }

    public int solution(String s) {
        int answer = 0;
        for (int i = 0; i < s.length(); i++) {
            String rotateString = rotate(s, i);
            if (check(rotateString)) {
                answer += 1;
            }
        }
        return answer;
    }
}
